Script to train xgboost trees

smguo opened the pull request:
This PR includes the following changes:
1. Add the script to train xgboost classifiers using information from multiple antigens.
2. Update README with usage of the script and `analyze_od` module.

smguo committed:
Add script to train xgboost trees; update README.md

jennyfolkesson commented on README.md:
```diff
@@ -9,14 +9,9 @@ The project automates estimation of antibody titers from data collected with ELI
-
-On a typical Winodws, Mac, or Linux computer:
+On a typical Windows, Mac, or Linux computer:
 * Create a conda environment: `conda create --name pysero python=3.7`
```
Just a suggestions: for simplicity you could also create a conda environment yaml file for creating the environment with a one liner. 
E.g. https://github.com/czbiohub/microDL/blob/master/conda_environment.yml

smguo replied:
Good suggestion! I added a conda environment yaml file

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    :param dataframe dtrain: training data with rows being samples and columns being features
+    :param list features: column names of features
+    :param str target: column name of the target
+    :param bool cv: if true, optimal number of estimators is determined by cross-validation
```
Spell it out as cross_valid or similar, cv is already used for opencv.

smguo replied:
Here I'm trying to be consistent with the scikit-learn classifier parameter convention. opencv is almost never passed as a function argument so I think this should be ok?

jennyfolkesson replied:
Who knows, you might be ok bug-wise... but since cv is used for a different purpose elsewhere in this repo one could argue that you should be consistent with pysero conventions first and other module conventions second.

smguo replied:
Good point on naming convention consistency throughout the repo. I've changed `cv` to `cross_valid`

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+                          metrics='auc', early_stopping_rounds=early_stopping_rounds, stratified=True)
+        model.set_params(n_estimators=cvresult.shape[0])
+        test_score = cvresult['test-auc-mean'].iloc[-1]
+        print('# of estimators: %d' % cvresult.shape[0])
```
Do you want to print this? Either way, add to logger too.

smguo replied:
added to logger

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    train_score = metrics.accuracy_score(dtrain[target].tolist(), train_yhat)
+    if cv:
+        score = test_score
+        print("AUC Score (Test): %f" % score)
```
Add to logger

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    :param list features: column names of features
+    :param str target: column name of the target
+    :param dict or list of dictionaries param_test: Dictionary with parameters names (string)
+        as keys and lists of
```
Weird line break?

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+        dictionaries, in which case the grids spanned by each dictionary
+        in the list are explored. This enables searching over any sequence
+        of parameter settings.
+    :param int, cross-validation generator or an iterable cv: cross-validation splitting strategy.
```
I feel this one could need a bit more of an explanation since there are so many options...

smguo replied:
Added reference to `GridSearchCV`

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+        in the list are explored. This enables searching over any sequence
+        of parameter settings.
+    :param int, cross-validation generator or an iterable cv: cross-validation splitting strategy.
+    :param int or None n_jobs: Number of jobs to run in parallel.
```
None means only one job? Same as 1 or different? Parallel as in multiprocessing?

smguo replied:
None is the same as 1 according to [joblib's documentation page](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html):

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    gsearch.fit(train[features], train[target])
+    model.set_params(**gsearch.best_params_)
+    means = gsearch.cv_results_['mean_test_score']
+    print(means)
```
Remove print

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    means = gsearch.cv_results_['mean_test_score']
+    print(means)
+    stds = gsearch.cv_results_['std_test_score']
+    for mean, std, params in zip(means, stds, gsearch.cv_results_['params']):
```
It seems like this function is three lines of code then a bunch of print statements... can some prints be moved to log instead?

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    return model
+
+def plot_xgb_fscore(model, output_dir, output_fname):
+    """plot xgboost feature importance"""
```
Add params.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+import pandas as pd
```
the interpretation module needs an __init__.py

smguo replied:
Added init.py

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+from sklearn.model_selection import GridSearchCV, GroupKFold, GroupShuffleSplit
+from xgboost.sklearn import XGBClassifier
+import xgboost as xgb
+#%%
```
For pep you need to add double spacing between functions.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+
+if __name__ == '__main__':
+    #%% set file paths and load OD table
+    data_dir = r'/Volumes/GoogleDrive/My Drive/ELISAarrayReader/manuscript/data_fig_4/output'
```
This assumes whoever runs it will have the same drive structure and data in it as you... is that what you want?

smguo replied:
Moved this to `example` within the repo

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+from matplotlib import pyplot as plt
+import unicodedata
+from interpretation.plotting import get_roc_df, roc_plot_grid, thr_plot_grid
+from interpretation.report_reader import slice_df, normalize_od, offset_od
```
Ideally imports should be sorted alphabetically with local imports saved for last.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+from sklearn import metrics
+from sklearn.model_selection import GridSearchCV, GroupKFold, GroupShuffleSplit
+from xgboost.sklearn import XGBClassifier
+import xgboost as xgb
```
You've already imported all of xgboost, so call the classifier using xgb.sklearn.XGBClassifier instead of importing again.

jennyfolkesson replied:
You need to add xgboost to requirements.
You also need to make sure that users have OpenMP installed, probably by doing a conda install.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    df_norm['antigen_col'] = df_norm['antigen_col'].map(str)
+    pysero_df_pivot = df_norm.copy()
+    pysero_df_pivot = pd.pivot_table(pysero_df_pivot, values='OD',
+                                 index=['plate ID', 'well_id', 'serum ID',
```
Weird line breaks - you could define index prior to function call for an easier read.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    slice_keys = [sera_roc_list, ['Diagnostic'], antigen_list]
+    slice_actions = ['drop', 'keep', 'keep']
+    fpr = 0.05
+    ci = 95
```
Constants should be defined at the top of the file with capitals.

smguo replied:
Not in this case because they are intended to be local variable

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    :param int or None n_jobs: Number of jobs to run in parallel.
+    :return object model: estimator object with optimal parameters.
+    """
+    gsearch = GridSearchCV(estimator=model, param_grid=param_test,
```
Perhaps move each parameter to new line for readability.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+    for param_test in param_tests:
+        clf = tune_cls_para(clf, train, features, target='target', param_test=param_test, cv=folds, n_jobs=-1)
+
+    #%% Train the classifier with optimal parameters
```
This comment just confused me because the optimal parameters are from tune_cls_para, or did you find optimal values for learning rate and number of estimators somehow as well?

smguo replied:
No just lower the learning rate and increase the number of rounds empirically. Clarified this in the comment now.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+                        'secondary dilution'])['OD'].mean()
+    roc_df = roc_df.reset_index()
+    roc_df = pd.concat([test, roc_df])
+    _ = roc_plot_grid(roc_df, fig_path, '_'.join(['ROC', suffix]), 'pdf', col_wrap=4, ci=ci, fpr=fpr, hue=hue)
```
Do you ever need to return something from a plot function? Maybe consider dropping the return?

smguo replied:
Yes the function also returns a dataframe

smguo replied:
Yes the function also returns a dataframe

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+        df.loc[:, 'antigen type'] = 'Diagnostic'
+    #%% transform the dataframe back to long format for plotting
+    test_keys = test.drop(features + ['target'], axis=1)
+    plot_xgb_fscore(clf, output_dir=fig_path, output_fname='xgb_feature_importance')
```
Maybe add some more comments in this function to make talk the user through what's happening.

jennyfolkesson commented on interpretation/train_classifier.py:
```diff
@@ -0,0 +1,219 @@
+                dpi=300, bbox_inches='tight')
+    plt.close()
+
+if __name__ == '__main__':
```
You should put most of your code in functions or classes, not in __main__. See e.g. 
https://realpython.com/python-main-function/

smguo replied:
I moved most of the code to the function `main()` now. This is intended to be more a script demonstrating how to build a classifier for pysero output rather a general utility for training classifier (or at least not at the moment). Perhaps we can refactor this later?

jennyfolkesson reviewed (changes requested):
I added quite a few comments, but I think the person reviewing will appreciate it if you make their lives easy for them. Happy to do a more thorough review once the script is cleaned up a bit!

smguo committed:
load input file from `examples`; restructure

smguo committed:
Merge branch 'master' into interprete_CLI

smguo committed:
Add logging

smguo committed:
Add logistic regression

smguo committed:
Add docstrings

smguo committed:
Add conda environment yaml

smguo commented:
@jennyfolkesson Sorry for the long delay in getting to this PR. I was working on other projects. I have addressed most of your comments. One thing I should have noted is this PR is intended to be more a script demonstrating how to build a classifier for pysero output rather a general utility for training classifier. We can do more refactor later if this approach turns out to be useful for other projects too. 

@lenafb It would be great if you can help review this PR too. Let me know if you have any question!

lenafb reviewed (approved):
This makes sense to me. One suggestion/clarification: should the name of the env conda_env.yml be multiSero?

jennyfolkesson commented:
@smguo glad to see that pysero is getting some love again. Just let me know if you want me to contribute in any way.
LGMT

smguo committed:
change `cv` to `cross_valid`

smguo committed:
change conda environment name

smguo committed:
added init.py to example
